File: backend/src/hypomnema/llm/claude.py
"""Claude (Anthropic) LLM client."""

import json
from typing import Any

from anthropic import AsyncAnthropic


class ClaudeLLMClient:
    DEFAULT_MODEL = "claude-sonnet-4-20250514"
    DEFAULT_MAX_TOKENS = 4096

    def __init__(self, api_key: str, *, model: str = "", max_tokens: int = 0) -> None:
        self._client = AsyncAnthropic(api_key=api_key)
        self._model = model or self.DEFAULT_MODEL
        self._max_tokens = max_tokens or self.DEFAULT_MAX_TOKENS

    async def complete(self, prompt: str, *, system: str = "") -> str:
        kwargs: dict[str, Any] = {
            "model": self._model,
            "max_tokens": self._max_tokens,
            "messages": [{"role": "user", "content": prompt}],
        }
        if system:
            kwargs["system"] = system
        response = await self._client.messages.create(**kwargs)
        block = response.content[0]
        if not hasattr(block, "text"):
            raise TypeError(f"Expected TextBlock, got {type(block)}")
        return str(block.text)

    async def complete_json(self, prompt: str, *, system: str = "") -> dict[str, Any]:
        text = await self.complete(prompt, system=system)
        return dict(json.loads(text))

File: server/internal/llm/claude.go
package llm

import (
	"context"
	"encoding/json"
	"fmt"
)

const claudeAPI = "https://api.anthropic.com/v1/messages"

type claude struct {
	model  string
	apiKey string
}

func newClaude(model, apiKey string) *claude {
	if model == "" {
		model = "claude-sonnet-4-20250514"
	}
	return &claude{model: model, apiKey: apiKey}
}

func (c *claude) Provider() string { return "claude" }

func (c *claude) Complete(ctx context.Context, prompt, system string) (string, error) {
	body := map[string]any{
		"model":      c.model,
		"max_tokens": 4096,
		"messages":   []map[string]string{{"role": "user", "content": prompt}},
	}
	if system != "" {
		body["system"] = system
	}

	respData, err := httpDo(ctx, "POST", claudeAPI, map[string]string{
		"x-api-key":         c.apiKey,
		"anthropic-version": "2023-06-01",
	}, body)
	if err != nil {
		return "", fmt.Errorf("claude: %w", err)
	}

	var resp struct {
		Content []struct {
			Text string `json:"text"`
		} `json:"content"`
	}
	if err := json.Unmarshal(respData, &resp); err != nil {
		return "", fmt.Errorf("claude parse: %w", err)
	}
	if len(resp.Content) == 0 {
		return "", fmt.Errorf("claude: empty response")
	}
	return resp.Content[0].Text, nil
}

func (c *claude) CompleteJSON(ctx context.Context, prompt, system string) (map[string]any, error) {
	text, err := c.Complete(ctx, prompt, system)
	if err != nil {
		return nil, err
	}
	return extractJSON(text)
}
